File: DotNet/Kyu6/SimpleFun165Withdraw.cs
// https://www.codewars.com/kata/simple-fun-number-165-withdraw/train/csharp

namespace DotNet.Kyu6;

public class SimpleFun165Withdraw
{
    public int[] Withdraw(int n)
    {
        for (int hundreds = n / 100; hundreds >= 0; hundreds--)
        {
            for (int fifties = n / 50; fifties >= 0; fifties--)
            {
                for (int twenties = n / 20; twenties >= 0; twenties--)
                {
                    if (n == hundreds * 100 + fifties * 50 + twenties * 20)
                    {
                        return [hundreds, fifties, twenties];
                    }
                }
            }
        }

        return [];
    }

    public static void Main()
    {
        SimpleFun165Withdraw kata = new SimpleFun165Withdraw();
        Console.WriteLine(string.Join(", ", kata.Withdraw(40))); // [0, 0, 2]
        Console.WriteLine(string.Join(", ", kata.Withdraw(250))); // [2, 1, 0]
        Console.WriteLine(string.Join(", ", kata.Withdraw(260))); // [2, 0, 3]
        Console.WriteLine(string.Join(", ", kata.Withdraw(230))); // [1, 1, 4]
        Console.WriteLine(string.Join(", ", kata.Withdraw(60))); // [0, 0, 3]
    }
}

File: Elixir/Kyu6/simple-fun-165-withdraw.ex
# https://www.codewars.com/kata/simple-fun-number-165-withdraw/train/elixir

defmodule ATM do
  def is_valid_combination?(amount, hundreds, fifties, twenties) do
    amount == hundreds * 100 + fifties * 50 + twenties * 20
  end

  def first_valid_combination(amount, max_hundreds, max_fifties, max_twenties) do
    for hundreds <- max_hundreds..0,
        fifties <- max_fifties..0,
        twenties <- 0..max_twenties do
      if is_valid_combination?(amount, hundreds, fifties, twenties) do
        throw([hundreds, fifties, twenties])
      end
    end
  end

  def withdraw(n) do
    max_hundreds = div(n, 100)
    max_fifties = div(n, 50)
    max_twenties = div(n, 20)

    try do
      first_valid_combination(n, max_hundreds, max_fifties, max_twenties)
    catch
      combination -> combination
    end
  end
end

IO.inspect(ATM.withdraw(40))
IO.inspect([0, 0, 2])
IO.inspect("-------")

IO.inspect(ATM.withdraw(60))
IO.inspect([0, 0, 3])
IO.inspect("-------")

IO.inspect(ATM.withdraw(250))
IO.inspect([2, 1, 0])
IO.inspect("-------")

IO.inspect(ATM.withdraw(260))
IO.inspect([2, 0, 3])
IO.inspect("-------")

IO.inspect(ATM.withdraw(370))
IO.inspect([3, 1, 1])
IO.inspect("-------")

IO.inspect(ATM.withdraw(10000))
IO.inspect([100, 0, 0])
IO.inspect("-------")
